Removing VersionedResourceTypes from get, check and put plans

clarafu opened the pull request:
## Changes proposed by this PR

This PR removes the need to include `VersionedResourceTypes` inside of the `get`, `check` and `put` plans. It is still necessary in the `task` plans because of the fact that the task config might not be available to the ATC during the time of the build planner due to the possibility of the task config being an external file (using `file: <file-path>`).

I've wanted to do this for a really long time because it always seemed weird how we always include the list of ALL the custom resource types in each step plan, even if the step is for a resource that does not use a custom resource type. In order to make this happen, we introduced a new field inside the step plans called `TypeImage`. This `TypeImage` struct includes a `GetPlan`, `CheckPlan`, `BaseType` and `Privileged`: 
- `GetPlan` : contains the get plan for fetching the image. Not nil when the resource step is using a custom resource type. 
- `CheckPlan`:  contains the check plan for grabbing the resource version for the image type. Not nil when the resource step is using a custom resource type.
- `BaseType`: contains the base resource type all the way down the dependency chain, so if you have a resource `foo` that has a parent custom resource type of `bar` and the `bar` resource type's parent is `git`, the `BaseType` field for the resource `foo` will contain `git` as the value.
- `Privileged` is used for indicating whether the parent resource type is privileged.

One of the reasons we needed the `VersionedResourceTypes` in the step plans was so that we could find or create resource configs or caches for the step. In order to remove the need for it, we are actually just directly using the resource cache returned back from fetching the image within the step (since the end result of why we needed the versioned resource types was to grab the resource cache of the parent). By doing this, we were able to remove the `Descriptor` layer of `ResourceConfigDescriptor` and `ResourceCacheDescriptor` (yay!) that was previously used to construct the resource config/cache using the versioned resource types. 

One other performance benefit is that the `ResourceTypes` query will no longer need to grab a version for each resource type when we fetch them from the database. So this query below is no longer there

https://github.com/concourse/concourse/blob/7fa86c1900311cd061203ee413b9c8113f8bf09c/atc/db/resource_type.go#L161-L167

That part of the query has actually been reported as one of the slowest queries for some of our users that most likely had a lot of custom resource types and versions.

**One important point to note**: There actually are behavioural changes in this PR, it is NOT a pure refactor. The behavioural change is that every time we run a step, we will actually run a check for the parent type. That check will still respect the checking interval/check every so that will ensure we don't check more times than the checking interval. This means that we will be removing the explicit checking for resource types in lidar. 

part of #6274 

- [x] Remove `container_id` from resource cache uses
- [x] Change checkfactory.ResourceTypes to return a map
- [x] Clean up/remove comments in atc.Plan
- [x] Web UI to use the new GetPlan and CheckPlan for images
- [x] Fix bug with setting `Interval` on check plans created in FetchImagePlan/ImageForType
       - Fixing this bug will result in fly execute builds to always run a check for any  custom resource types that are used as an image in the build. This is because we don't really have an easy way to propagate the default checking interval to fly for when we construct the check plan, also I think it makes sense for fly executed builds to run a check for any images used.
- [x] Fix bug with custom types acquiring checking lock. (can exit early if we check the now - lastcheck time is less than interval, and while we are waiting on the lock we can also periodically check lastcheck time so we can exit early if it hits)
- [ ] Try to consolidate merging of source defaults
- [ ] Try to refactor the check plan creations into one single place (rather than right now it can be created by db objects or the atc ResourceType object) 

## Release Note

* This PR is mainly a refactor but there is a behavioural change that comes along with it. If a resource uses a custom resource type, its `Put`, `Check`, `Get` and `Task` steps will now always create a check for its parent custom resource type. This check will still respect the resource checking interval and the `check_every` of the resource type. Because of this new feature, we no longer need to explicitly check the custom resource types in `lidar`. 

[imperative mood]: https://chris.beams.io/posts/git-commit/#imperative

Commit:
atc: wip planning atc.Plan structure

- statically generate nested build plans for fetching the image for a
get/put step
- reference the imagespec and version of a step from the outputs of a
previous check/get step

TODO:
- need to watch out for taskconfig file

[#6274]

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Clara Fu <cfu@pivotal.io>

clarafu committed:
B: generate static plans for image sub steps

[#6274]

This is a squashed commit that included a TON of commits so I'm sorry
for the terrible commit message!

Previously, we pushed the creation of image check and get plans into the
planner but did it through individual steps wrapped together through on
successes. This brought some complications such as needing to change how
it is rendered in the web UI and also made it difficult to read the build
plan. This commit keeps the creation of the image check and gets in the
planner but rather than running them separately and tieing the step and
it's image check/get through the ImageFrom field, we inline the image
plans directly in the step itself. Because of this, all FetchImage
within each step that needs an image only needs to run the
preconstructed plan and retrieve the results from the run state.

There will be an ImageGetPlan and ImageCheckPlan added to each step plan
that requires an image. Each step will run these constructed plans,
similar to how it works now but only difference is it no longer needs to
create the plans.

The logic for the current FetchImage needs to be moved into the task
step since it still needs to create substeps for dynamically fetched
task configs.

Image check steps store versions in the runstate, which is then grabbed
by the get step and used. We modified the version grabbed from the get
step to not be specific to the put step, since we are now using it for
check steps as well. The planID specified by ImageSpecFrom actually
stored more results than just the ImageSpec, so let's make the name more
accurate.

The Check step stored just the version, but the Put step stored the
version and the metadata. We couldn't find a use for the
`VersionResult.Metadata` anywhere so let's only store the Version
instead

- add test coverage, most of the image spec generation logic will be
tested in the get step tests
- it would be painful to maintain different codepaths for when the
task config is known vs when it's dynamically fetched

TODO:
- pass the varsourceconfig through the scheduler into the build plan
- figure out why we're storing runtime.VersionResult, do we actually use
the VersionResult.Metadata in any subsequent operations?
- Because we removed generating new scopes in FetchImage, it's now
possible for artifacts generated by get step to collide with image get
steps. The solution we're thinking of is to split the artifact
repositories into 3, resource, resourceType, and anonymousResource. This
way we can rely on the guarantees provided by ConfigValidate (i.e. no
duplicate resource or resourceType names) and flatten the run state
without worrying about if a get for a resource type named "foo" will
collide with a resource named "foo".

Signed-off-by: Clara Fu <cfu@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Esteban Foronda <eforonda@vmware.com>
Co-authored-by: Aidan Oldershaw <aoldershaw@pivotal.io>

[#6274]

clarafu committed:
atc: add image check/get to public plans

Added the ImageGetPlan and ImageCheckPlan to all the steps that use an
image and also added the Public method to the GetVarPlan. Filled in a
test for the Public method on LoadVarPlan.

[#6274]

Signed-off-by: Clara Fu <cfu@pivotal.io>
Co-authored-by: Aidan Oldershaw <aoldershaw@pivotal.io>

clarafu committed:
S: remove builds -> engine dependency

by moving the interfaces to where they're actually used. this will help
prevent import cycles as we intend to use parts of the planner in
atc/exec

Signed-off-by: Clara Fu <cfu@pivotal.io>
Co-authored-by: Aidan Oldershaw <aoldershaw@pivotal.io>

clarafu committed:
wip: manual plan id generation for image get/check

plan IDs for image fetching are no longer sequentially generated - the
get plan has "/image-get" appended, and the check plan has
"/image-check" appended (to the parent plan's ID)

this commit also rewrites the FetchImage method to use the pre-computed
plans. the TaskStepDelegate has a special FetchImage method that
generates a new plan dynamically (as was the case before)

Also adds Privileged to resource step plans since we will be removing
VersionedResourceTypes from the resource step plans, we need to indicate
whether these steps are for a privileged resource type

[#6274]

Signed-off-by: Clara Fu <cfu@pivotal.io>
Co-authored-by: Aidan Oldershaw <aoldershaw@pivotal.io>

clarafu committed:
wip: store base resource type on plan

Signed-off-by: Clara Fu <cfu@pivotal.io>
Co-authored-by: Aidan Oldershaw <aoldershaw@pivotal.io>

clarafu committed:
backfill engine tests

After changing the behaviour of the FetchImage methods within the build
step delegate and the task delegate, this commit backfills some tests
for those newly modified methods. It also fixed some broken tests after
the merge with master.

There were some exec test failures from the merge with master and also
some added tests from the image fetch/check refactor.

[#6274]

Co-authored-by: Rui Yang <ryang@pivotal.io>
Signed-off-by: Clara Fu <fclara@vmware.com>
Signed-off-by: Bohan Chen <bochen@pivotal.io>

Commit:
find or create resource config and cache without descriptor

Removed the resource config and cache descriptors! We were able to do
that because we are now passing in the resource cache for the custom
type (if it uses one) to the FindOrCreateResourceConfig or
FindOrCreateResourceCache method, and with the resource cache we no
longer require the descriptor to tell us what its chain of parent types
are. By doing this, we can remove versioned resource types from the step
plans (such as get, check and put), we need to remove its usage from the
steps itself.

Currently FindCheckContainer is broken because it also used
FindOrCreateResourceConfig, but now that it takes in a resource cache we
dont have a resource cache to pass into it from the FindCheckContainer
method. We will be removing the need to get a resource config at all in
the future for FindCheckContainer so we can keep it broken for now.

Lastly, since we no longer have a ResourceCacheDescriptor, we don't need
the name of the resource cache object to be UsedResourceCache. (yay)

[#6274]

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Clara Fu <fclara@vmware.com>
Co-authored-by: Alex Suraci <suraci.alex@gmail.com>

clarafu committed:
atc: remove check planner object

This is a squash of many commits so Im sorry if this commit is pretty
big :D

- Originally, a resource or resource type db object would be used to
create a check plan. During this refactor, I changed it so that the
check plan would be created through a Check Planner within the builds
package. The reasoning was because I wanted all the logic for creating
build plans to be in one place because we were adding more complex
logic with building image plans. After discussing this with Alex, we
decided it might be better to move the creation of plans back to the
resource and resource type object. This is because the check plan that
needs to be created for a resource and resource type is slightly
different, and also it is better practice to keep a small interface for
a narrowly scoped method (for ex, creating an image plan) and to use
that small interface in any part of the codebase (rather than if the
interface was a check planner which holds a bigger scope).

- We also refactored the added image fields
(ImageGetPlan, ImageCheckPlan and BaseType) into a struct labelled as
"image" in the plans. We also moved Privileged into the image struct
because the privileged boolean is related to the parent custom type.

- The versioned resource types will now contain the method for
constructing an image plan. This is because it seemed to be the place
that makes the most sense for where it belongs, since the image plan is
for fetching the image of a custom resource type.

- In the TryCreateCheck method, if the list of resource types is nil we
need to convert it to an empty list. This is because we are using the
versioned resource types as an object that conforms to the ImagePlanner
interface, and if it is nil then the code that calls a method off of
the ImagePlanner interface will panic.

- In the Image plan of a step plans (get, check, put and task), whenever
we need to fetch an image for a custom type we will now always create a
check step for the custom type. This won't have any performance
regressions because the check will only run if it has been less than the
checking interval, so we don't have to worry about a hoard of checks
being run for custom types. This will allow us to remove the need for
checking resource types explicitly in lidar.

[#6274]

Signed-off-by: Clara Fu <fclara@vmware.com>
Co-authored-by: Alex Suraci <suraci.alex@gmail.com>

clarafu committed:
atc: remove atc.VersionedResourceTypes

Since we no longer use the version off of the VersionedResourceTypes
(because we always create a check whenever a custom type is used) inside
a build, we can remove this type entirely and replace all usages with
ResourceTypes. This commit replaces all the places where
VersionedResourceTypes was used within the ATC. We still need to convert
go-concourse and fly to no longer use it.

This might have some performance benefits because we no longer need to
query for a version for each resource type when we fetch the resource
types from the database.

Also removed versioned resource types from fly and go-concourse. A few
major changes within this commit:

- go-concourse: change VersionedResourceTypes to ResourceTypes. The
resource types returned from that method will not longer contain the
versions for each resource type, which should be okay since nothing used
the version.

- fly execute will construct a TypeImage within the GetPlans for each
input. This way fly execute will be able to construct the correct build
plan for each input.

- fly check-resource and check-resource-type no longer run an individual fly
check-resource-type for its parent custom type because a resource check
step will now result in its parent being checked as the image plan. This
means we dont need to run an extra check to ensure the parent type is
checked. One problem that came about this was that when a user triggers
a new manually triggered build, this will result in the resources AND
its parent custom type forcing a check to happen. If they had 5
resources that all use the same custom type, it will force 5 checks on
the resources and then also 5 checks on the SAME custom type, compared
to before where it would have only forced 5 checks on the resources. So
in order to fix this extra performance burden, we made it so that there
will be a SkipInterval boolean on the CheckPlan and only if that is set
to true will the check be forced to skip its interval. And this will
only be set to true from fly check-resources or check-resource-types
where shallow is false. Every other case, like check webhooks or
manually triggered builds, that creates resource checks will not force a
check so SkipInterval will be false.
[#6274]

Signed-off-by: Clara Fu <fclara@vmware.com>
Co-authored-by: Alex Suraci <suraci.alex@gmail.com>
Co-authored-by: Aidan Oldershaw <aoldershaw@pivotal.io>

clarafu committed:
wip: fixup after rebasing

Theres a bunch of small errors across the codebase after rebasing and
shifting around the commits with the plan refactor + get var step. Just
fixing that up in this commit and removing all traces of get var step!

[#6274]

Signed-off-by: Clara Fu <fclara@vmware.com>
Co-authored-by: Taylor Silva <tsilva@pivotal.io>

clarafu committed:
atc/lidar: remove resource-type checking

resource types are now checked within a build's plan

Signed-off-by: Clara Fu <fclara@vmware.com>
Co-authored-by: Taylor Silva <tsilva@pivotal.io>

clarafu committed:
remove container_id from resource cache uses

It is not used anymore after checks as builds, so we just removed it!

[#6274]

Signed-off-by: Clara Fu <fclara@vmware.com>
Co-authored-by: Rui Yang <ryang@pivotal.io>

aoldershaw commented on atc/db/team.go:
```diff
@@ -955,33 +955,37 @@ func (t *team) FindCheckContainers(logger lager.Logger, pipelineRef atc.Pipeline
+	// find any builds that have resource_id = resource.id
+	// Find any containers that have build_id ^
+
+	// XXX: THIS DOES NOT WORK!
```
Leaving this as a reminder since it's not listed in the TODOs.

p.s. is there a reason we need to `FindOrCreateResourceConfig` here? Could we instead use the condition:
```
sq.Eq{"rccs.resource_config_id": resource.ResourceConfigID()}
```

I guess one implication of that is that we'd need to have called `PointToCheckedConfig` at least once in the check delegate (i.e. at least one check had to have completed, either successfully or unsuccessfully).

aoldershaw commented on atc/engine/step_factory.go:
```diff
@@ -20,6 +21,7 @@ type coreStepFactory struct {
 	resourceFactory       resource.ResourceFactory
 	teamFactory           db.TeamFactory
 	buildFactory          db.BuildFactory
+	lockFactory           lock.LockFactory
```
Can we get rid of this?

aoldershaw commented on atc/exec/build/variables.go:
```diff
@@ -0,0 +1,76 @@
+	lock        sync.RWMutex
+}
+
+func NewVariables(tracker *vars.Tracker) *Variables {
```
Was this part of the get var work? Looks like it's unused

aoldershaw commented on atc/public_plan.go:
```diff
@@ -14,6 +18,7 @@ func (plan Plan) Public() *json.RawMessage {
 		Check          *json.RawMessage `json:"check,omitempty"`
 		Task           *json.RawMessage `json:"task,omitempty"`
 		SetPipeline    *json.RawMessage `json:"set_pipeline,omitempty"`
+		GetVar         *json.RawMessage `json:"get_var,omitempty"`
```
We can get rid of this for now

aoldershaw commented on foo:
```diff
@@ -0,0 +1,15 @@
+task:
```
Can delete this

aoldershaw reviewed (changes requested):
Just did a quick skim, have some minor comments so far.

Also - I think we'll need some UI changes, as this seems to break the rendering of "image fetching" for check/get/put plans. Since we no longer emit the check/get plans in the build events for these types, we never end up rendering the cog icon. We'll need to decode the fetch image plans here: https://github.com/concourse/concourse/blob/d23db1bb0acb440238d8c8cd41ac81edb24aba0e/web/elm/src/Concourse.elm#L682-L687

aoldershaw commented on atc/engine/build_step_delegate.go:
```diff
@@ -225,139 +211,71 @@ func (delegate *buildStepDelegate) Errored(logger lager.Logger, message string)
-		return worker.ImageSpec{}, fmt.Errorf("get did not return a cache")
+	var result exec.GetResult
+	if !fetchState.Result(getPlan.ID, &result) {
+		return worker.ImageSpec{}, nil, fmt.Errorf("get did not return a result")
```
Haven't looked into it too much, but I seem to be hitting this case pretty frequently
<img width="499" alt="Screen Shot 2021-06-22 at 11 33 07 AM" src="https://user-images.githubusercontent.com/26583442/122954655-a048cc00-d34d-11eb-9d09-5b51d50566e7.png">

clarafu committed:
check factory resource types grouped by pipelineID

The ResourceTypes method on the check factory will now return back a map
of pipelineID to resource types in order to easily pass in the correct
list of resource types for a specific pipeline. This fixes a bug in the
CheckPlan method, where it expects the list of resource types to be
already condensed to a list of resource types within the specific
pipeline.

[#6274]

Signed-off-by: Clara Fu <fclara@vmware.com>
Co-authored-by: Rui Yang <ryang@pivotal.io>

clarafu committed:
remove some unnecessary comments during planning

[#6274]

Signed-off-by: Clara Fu <fclara@vmware.com>
Co-authored-by: Rui Yang <ryang@pivotal.io>

clarafu commented:
Oh ya I never did the web UI portion of this 🤭  Will take a look at that now haha

clarafu committed:
fix db tests after rebase

[#6274]

Signed-off-by: Clara Fu <fclara@vmware.com>
Co-authored-by: Aidan Oldershaw <aoldershaw@pivotal.io>

clarafu committed:
web: show image steps from build plan

Previously, we were showing the image steps through build events that
happened after the initial build page was loaded. This is changed in
that we will now send the image steps within the initial build plan.
This affects the get, check and put steps while the task step will
continue to show image steps through build events.

[#6274]

Signed-off-by: Clara Fu <fclara@vmware.com>
Co-authored-by: Aidan Oldershaw <aoldershaw@pivotal.io>

clarafu commented on atc/db/resource.go:
```diff
@@ -255,22 +251,35 @@ func (r *resource) setResourceConfigScopeInTransaction(tx Tx, scope ResourceConf
 }
 
+// CreateBuild creates a check build for this resource. It returns back the
+// build that was created, whether or not it was successful and any errors.
```
Great suggestion!! :D

clarafu committed:
atc: make changes from feedback

Mostly composed of minor changes from Aidans feedback but one bug that
we did catch was that the get step fetching from local cache stored a
resource cache instead of a GetResult (which caused the get step to fail
when we fetch from the local cache)

[#6274]

Signed-off-by: Clara Fu <fclara@vmware.com>
Co-authored-by: Aidan Oldershaw <aoldershaw@pivotal.io>

clarafu committed:
S: refactor FetchImagePlan construction

We had some duplicated logic between builds.FetchImagePlan and
ResourceTypes.ImageForType in the construction of image fetching plans.
Now, the two share a single codepath.

Signed-off-by: Clara Fu <fclara@vmware.com>
Co-authored-by: Aidan Oldershaw <aoldershaw@pivotal.io>

clarafu committed:
atc: fix FindCheckContainer to not construct resource config

FindCheckContainer will now just read the resource config id off of the
resources table rather than constructing it. It needs the resource
config id in order to find the check containers associated to the
resource. This should result in the same behaviour because theoretically
the only benefit to constructing vs using the existing one in the db is
that it will always be up to date but we actually null out the resource
config id when the resource configuration is updated, so using the value
in the table should never result in the wrong resource config id being
used.

[#6274]

Signed-off-by: Clara Fu <fclara@vmware.com>
Co-authored-by: Aidan Oldershaw <aoldershaw@pivotal.io>

clarafu committed:
atc: fixup tests from previous changes

minor test fixes

[#6274]

Signed-off-by: Clara Fu <fclara@vmware.com>
Co-authored-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw commented:
@clarafu just a reminder about the testflight failures - it seems like the issue is with the tests themselves rather than this PR breaking anything, but I'm not 100% sure

clarafu committed:
fix go-concourse test

adds shallow argument to check prototypes.

Signed-off-by: Clara Fu <fclara@vmware.com>

clarafu commented:
Hm I think the last testflight failures are caused by us no longer printing the `checking <resource> in build <number>` for each recursive resource/resource-type that we check because we no longer recursively check through fly. This is what you would used to see in fly

```
checking test/mock-resource-parent in build 5
selected worker: 0ed52d516c6f
succeeded

checking test/mock-resource-child in build 6
INFO: found resource cache from local cache

selected worker: 0ed52d516c6f
succeeded

checking test/mock-resource-grandchild in build 7
INFO: found resource cache from local cache

INFO: found resource cache from local cache

selected worker: 0ed52d516c6f
succeeded

checking test/recursive-custom-resource in build 8
INFO: found resource cache from local cache

INFO: found resource cache from local cache

INFO: found resource cache from local cache

selected worker: 0ed52d516c6f
succeeded
```

And now it is 

```
checking tf-pipeline-1-a819454e-e2a1-44be-65ed-cd1643f87e98/recursive-custom-resource in build 192
selected worker: 8f0763df7c8c
INFO: found resource cache from local cache

selected worker: 8f0763df7c8c
INFO: found resource cache from local cache

selected worker: 8f0763df7c8c
INFO[0000] fetching version: child-version              
INFO[0000] mirroring self image                          via_params=false
selected worker: 8f0763df7c8c
INFO: found resource cache from local cache

INFO: found resource cache from local cache

INFO: found resource cache from local cache

selected worker: 8f0763df7c8c
INFO[0000] fetching version: grandchild-version         
INFO[0000] mirroring self image                          via_params=false
selected worker: 8f0763df7c8c
succeeded
```

My dilemma is that we don't want to just print out the names of the resources in the check step output and since we no longer use fly to recursively check, there won't be the messages of what exactly it is currently checking in the fly output.

aoldershaw commented:
@clarafu An imperfect solution could just be to count the number of `selected worker` events we receive

clarafu committed:
fix testflight failures

Some of the tests were testing that the resource type version does not
exist is errored but we no longer throw that error anymore because we
run a check everytime for any parent resource types. The other test that
was failing was testing that the resource/resource type gets printed
when it is checked recursively but that also does not exist anymore
because we are no longer checking recursively through fly (which is the
thing that printed that message).

[#6274]

Signed-off-by: Clara Fu <fclara@vmware.com>

clarafu commented:
It's weird, do the `selected worker` messages ever get lost? Sometimes when I run the tests it only prints a few of the `selected worker` messages but when I run the fly check resource command myself I always get 4 messages.

Also theres a test with resource type defaults that is failing so I'm looking into that. 🤔

aoldershaw commented:
It could be that a worker is only selected when a container actually runs, so if it's within the check interval it doesn't?

clarafu commented:
Don't we always run the check step if it is manually triggered?

clarafu commented:
So weird too, the defaults test is failing because it does not print out `"defaulted"` as a log for the resource type default 

```
checking tf-pipeline-1-e6a8c14d-6bfe-49b8-7390-f9af3257dfb8/some-resource in build 87
selected worker: 10b09931791e
INFO: found resource cache from local cache

succeeded
```

But when I run the check resource manually using the same pipeline, the log does show up. It's like the output gets cut out from the testflight tests..

```
checking test2/some-resource in build 98
selected worker: 10b09931791e
INFO: found resource cache from local cache

selected worker: 10b09931791e
INFO[0000] defaulted
succeeded
```

clarafu committed:
bump migrations

Signed-off-by: Clara Fu <fclara@vmware.com>

clarafu commented:
I've been trying to reproduce the testflight failure on my local machine but I've only got it to happen once after I repulled latest docker `concourse/dev` image and also added the two extra docker compose containerd override flags. But after it happened that one time it has continued to work properly after that, making me think it had nothing to do with those specific things that I changed and maybe the state of the workers/web when I was running the check resource commands? I was building docker compose and running the commands very shortly after it was up.

clarafu commented:
Ok I have a theory, I think this might be caused by 

https://github.com/concourse/concourse/blob/c867790fb5b924930ccb6d6f624a94846e414255/atc/engine/check_delegate.go#L148

where we avoid running a check if the build creation time is before the last check start time. I printed out the values when the test fails and this is what I get `shouldRun: false , lastCheck succeeded: true , create time after start boolean:  false , build create time:  2021-07-15 14:50:36.323729 +0000 +0000 , last check start time:  2021-07-15 14:50:37.155397 +0000 +0000`. I'm guessing this is only a problem now because now we run all the recursive parent resource type checks within the same build as the check for the resource, so the last check start time will always be later than the build creation time for any parent resource types. Rather than before we would create separate builds for each check we run for the recursive parent resource types.

aoldershaw commented:
@clarafu interesting, good find! I don't think I fully understand why this would occur, though. Why would the last check for the resource type's scope be after the build was created if we're in the process of determining whether we should run the check for the resource type?

Is this because other testflight tests are running with a `mock` resource with the exact same `resource_config` (and thus share a `resource_config_scope`)? i.e. there are other resources/resource types being checked concurrently with the same resource config as https://github.com/concourse/concourse/blob/fa6ec1a1b6bde3c25689abfcc1bceaeffdd1061f/testflight/fixtures/resource-type-defaults.yml#L3-L6

Or is something else going on?

clarafu commented:
Ya I actually thinnk I'm thinking about this wrong lol I think its the resource check that doesn't run.. right? Because that should be the second check that runs and the one that logs the `defaulted` output. I'm delirious!

aoldershaw commented:
Ohhhh interesting. When we unpause a pipeline, we kick off lidar immediately, so it's not that surprising that there would be an "interval" check for the resource right after the manual check starts.

I wonder why this wasn't a problem before...could it just be that because we now have to check the parent type, it just takes longer from after we create the check to actually run it?

clarafu commented:
Ohh yeah!! Ur probably right, we used to error when the resource type did not have a version.

aoldershaw commented:
Yeah, maybe we just get rid of the explicit fly check-resource in the test

clarafu commented:
Hm I see, we would need to get the build output from the last check then? Through `fly builds` then `fly watch`?

aoldershaw commented:
Oh i just mean don't test it for the check at all. We have unit tests for it, and if it works for the other step types than it's probably fine 😁  don't think it's worth blocking this PR, at least.

clarafu commented:
Ohh i see okay ill just do that then heh

clarafu committed:
testflight: remove check in resource type defaults test

The test for checking if the defaults are propagated to resource sources
for check step was flaking because now that we no longer error when the
parent resource type does not have a version, the initial interval check
that happens once we unpause the pipeline goes through and updates the
last check start time. This becomes a problem when that time that it
updates the last check start time is after the time that we also manuall
trigger a check resource for the test. If the last check start time is
after the time that we manually check the resource, the manual check
will not run (https://github.com/concourse/concourse/pull/7239). This is
as intended so I think we can just remove checking for the defaults for
a check step as it is tested in unit tests and also we will be checking
for the defaults in the get and put steps.

[#6274]

Signed-off-by: Clara Fu <fclara@vmware.com>

clarafu committed:
do not delete scopes with same resource config id

There is a chance that we are deleting the same resource config scope
(meaning equal resource_id and resource_config_id) when there are
multiple FindOrCreateScope calls running at the same time. This is
because both calls can not find a resource scope for a resource_id and
resource_config_id, but then both try to create one by going into the
codepath that will first delete outdated scopes and then insert a new
one. Rather than deleting outdated scopes, the second run can end up
deleting the scope created by the first run and the first run ends up
using a scope that does not exist.

[#6274]

Signed-off-by: Clara Fu <fclara@vmware.com>
Co-authored-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw commented on atc/db/resource_type.go:
```diff
@@ -48,16 +48,15 @@ type ResourceType interface {
 
 type ResourceTypes []ResourceType
 
+// XXX: TODO Can remove the bool
```
Meh, not sure if it's necessary. Returning the `bool` at least forces callers to deal with the fact that there may be no parent - it's possible to forget to check `!= nil`

clarafu commented on atc/lidar/scanner.go:
```diff
@@ -107,7 +90,15 @@ func (s *scanner) check(ctx context.Context, checkable db.Checkable, resourceTyp
 	}
 
-	_, created, err := s.checkFactory.TryCreateCheck(lagerctx.NewContext(spanCtx, logger), checkable, resourceTypes, version, false)
+	// We don't want to pass in a nil list of resource types here if the pipeline
+	// does not have any resource types because we will be calling methods off of
+	// the db.ResourceTypes interface in the TryCreateCheck method
+	var resourceTypes db.ResourceTypes
+	if rts, found := resourceTypesMap[checkable.PipelineID()]; found {
+		resourceTypes = rts
+	}
+
```
Oh sorry I meant a nil `db.ResourceTypes` rather than passing a `nil` to the `TryCreateCheck` method. Does that make sense? D:

aoldershaw replied:
Still a bit confused tbh. `db.ResourceTypes` is just a `[]db.ResourceType` (not an interface), so `nil` should be fine, right? I don't see how this code is different from just:

```go
resourceTypes := resourceTypesMap[checkable.PipelineID()]
```

...since if it's not found, it'll just take the zero-value (a `nil` slice) anyway

clarafu replied:
Ohh my bad I thought that if you don't find a key in a map it will return a `nil` and not a nil slice. But I'm wrong! Looks like if you don't find a key in a map, it just returns the default nil value which is a nil slice of that type. I just wanted to make sure that it was never a real `nil` value being passed into `TryCreateCheck` because the `db.ResourceTypes` is being used as an interface in that method. Thanks for pointing that out haha

clarafu committed:
atc: add InitializeCheck build event

Check's will always emit an initialize event now. This made it easier to
test in testflight.

Signed-off-by: Clara Fu <fclara@vmware.com>
Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Taylor Silva <tsilva@pivotal.io>

aoldershaw reviewed (commented):
Leaving this review to say the last commit looks good! Happy to merge after we clean up the leftover stuff I mentioned in my last review

clarafu committed:
remove unneccessary comments from pr review

Signed-off-by: Clara Fu <fclara@vmware.com>

clarafu committed:
fix multiple resource type containers

This will fix a bug where there were multiple check containers for
resource types because the check plan did not have the ResourceType
field set (which is the thing that makes sure we acquire a lock for
checking that resource type).

[#6274]

Signed-off-by: Clara Fu <fclara@vmware.com>
Co-authored-by: Aidan Oldershaw <aoldershaw@pivotal.io>

clarafu committed:
remove image check/get container id

The image check/get container owner is no longer being used because of
the refactor done by alex a while ago where image containers are no
longer a one off thing and are proper get/check containers owned by a
build.

Signed-off-by: Clara Fu <fclara@vmware.com>
Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

evanchaoli commented:
@clarafu @aoldershaw This PR looks like a "merge disaster" to #7208. If this one is merged first, then I would have to spend a lot of time understanding code changes of this PR. So can we shamelessly request to merge #7208 first?

aoldershaw commented:
@evanchaoli yeah, it'll be a pain resolving all the conflicts, but that'll be the case regardless of which gets merged first. One benefit to getting this merged first is that it'll simplify #7208 somewhat since we no longer have to worry about periodic checks for resource types, so the "in memory build" type only has to deal with resources.

That said, I get your concern about needing to understand all of these changes, too. If we do end up merging this PR first, I'm more than happy to handle the merge conflicts, and can send you a patch of the merge commit that you can apply to your branch.

Does that sound reasonable to you?

clarafu committed:
atc: add interval to image check plans

We forgot to add the interval to the image check plans and this caused a
bug where image checks always ran. The interval on the check plan is
used to determine if we need to run if now - the last check time >
interval. If it has been less than the interval then we should not run,
this is so that we don't run a check for a custom type more than once
per interval.

Previously, we did not need to add the interval to the image check plans
because we only constructed an image check if the resource type did not
have a version (which should be only when it first runs in a pipeline).
Now that we always construct the image check, we need the interval to
limit the amount of times the check actually runs.

One thing to note is that image checks from fly execute builds will
always run. Previously, we did not always run a check if the image
alrady had a version but now that we don't have an interval set on the
plans constructed from fly execute, it will always run a check.

[#6274]

Co-authored-by: Esteban Foronda <eforonda@vmware.com>
Signed-off-by: Clara Fu <fclara@vmware.com>

clarafu committed:
atc: do not acquire checking lock if we do not need to run

Previously, we were first acquiring the checking lock and then seeing if
we need to run the check. This means that even if we know that we do not
need to run, it still holds the lock until the step completes. This
became a problem when we are creating checks for images and since a
custom type can be used by many resources across many jobs/pipelines,
the locking was causing some checks to take super long (15 minutes) to
acquire the lock.

So rather than acquiring the lock and then seeing if we need to run,
this commit changes it so that it will check to see if it needs to run
and only if it needs to run will it acquire the lock. It will do this
periodically every loop (1 second). I have slight concerns over the
amount of queries that will be run against the db because everytime it
loops for every resource/resource type/prototype, it will run a query to
fetch the last check times.

[#6274]

Signed-off-by: Clara Fu <fclara@vmware.com>

clarafu committed:
fix some tests

Signed-off-by: Clara Fu <fclara@vmware.com>

clarafu commented on atc/engine/check_delegate.go:
```diff
@@ -107,6 +121,29 @@ func (d *checkDelegate) WaitToRun(ctx context.Context, scope db.ResourceConfigSc
 	var lock lock.Lock = lock.NoopLock{}
 	if d.plan.IsPeriodic() {
 		for {
+			lastCheck, err := scope.LastCheck()
```
Please double check my logic here 😟  @aoldershaw

clarafu commented on atc/engine/check_delegate.go:
```diff
@@ -107,6 +121,29 @@ func (d *checkDelegate) WaitToRun(ctx context.Context, scope db.ResourceConfigSc
+				return nil, false, err
+			}
+
+			if d.build.IsManuallyTriggered() { // if the check was manually triggered
```
Please double check my logic here 😟  @aoldershaw, I changed the if statements from being setting `shouldRun` to true, to returning false. 

This is what I converted it from:
https://github.com/concourse/concourse/blob/b047d257f2531c119e15d627e3799d9d931bad03/atc/engine/check_delegate.go#L136-L152

evanchaoli replied:
Why move acquire lock to be after interval check? I'm not sure if the logic change is correct without a deep thinking, but if this optimization is not related to this PR, I would suggest divide this change to a separate PR, so that people can put more attention on it.

clarafu replied:
This actually is related because it fixes a bug that this PR introduced, I wrote a summary of it in the commit that adds it here https://github.com/concourse/concourse/pull/7176/commits/28982bddba3b1433ce4e345eecc0a56d7dda6aa0. Let me know if you have any more questions!

aoldershaw commented on atc/engine/check_delegate_test.go:
```diff
@@ -319,6 +329,29 @@ var _ = Describe("CheckDelegate", func() {
 					})
 				})
+
+				Context("when the last check gets updated after the first loop", func() {
```
What is this intended to test? I was expecting it to test that if it can't acquire the lock initially, it'll keep re-checking the `LastCheck` periodically, but it's really just testing the same thing as `when the interval has not elapsed since the last check`

clarafu replied:
Yaaaa right, it is exactly like that test. I don't know what I was thinking when I added this test haha 😆  

I think maybe I wanted to test that the last check value gets rechecked everytime it loops (but ofc I didn't write the test properly).

evanchaoli commented on atc/api/resourceserver/check.go:
```diff
@@ -55,6 +55,7 @@ func (s *Server) CheckResource(dbPipeline db.Pipeline) http.Handler {
 			dbResourceTypes,
 			reqBody.From,
 			true,
+			!reqBody.Shallow,
```
"Shallow" means check the resource itself. In that case, we suppose to not check resource types, then skipIntervalRecursively should to true. Not sure if my understanding is correct or not.

evanchaoli commented on atc/builds/tracker.go:
```diff
@@ -6,14 +6,27 @@ import (
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -generate
+
+//go:generate counterfeiter . Engine
+type Engine interface {
```
Looks like redefine `Engine` here just for defining a no-return `Runnable`. I feel not worthy doing so. If we want to silently ignore error returned from a run, `engine` may just not checking return value of `Run`. Adding a duplicate `Engine` and `Runnable` adds some confusion.

aoldershaw replied:
At one point, I think it may have been an import cycle - but even with no import cycle, the [Go authors recommend](https://github.com/golang/go/wiki/CodeReviewComments#interfaces) defining interfaces where they're used, rather than where they're implemented:

> Go interfaces generally belong in the package that uses values of the interface type, not the package that implements those values. The implementing package should return concrete (usually pointer or struct) types: that way, new methods can be added to implementations without requiring extensive refactoring.

As you've no doubt noticed, we do a bad job following that advice, and honestly it's caused some pain 😁  but this tends to be a more sane way of defining dependencies.

clarafu replied:
I actually would argue that it makes more sense to have it in the build tracker, this is because I found that the best practice is to create the interfaces where it is used rather where it is implemented. This makes it so that we can have more small and condensed interfaces while the implementation of the objects that conform to the interface be less tied to that interface. The interface should be directly tied to only what that specific component needs out of the object.

Throughout our codebase, we have interfaces really closely tied to the objects that implement them (really good example is the db package). But the only reason why we do that is to be able to test easily using fakes. I think if we can avoid tying the interface with the object, that would be best. What do you think about this?

evanchaoli replied:
Yeah, make sense.

evanchaoli commented on atc/config.go:
```diff
@@ -296,6 +296,104 @@ func (types ResourceTypes) Without(name string) ResourceTypes {
+
+func FetchImagePlan(planID PlanID, image ImageResource, resourceTypes ResourceTypes, stepTags Tags, skipInterval bool, checkEvery *CheckEvery) (Plan, *Plan) {
+	// If resource type is a custom type, recurse in order to resolve nested resource types
+	getPlanID := planID + "/image-get"
```
If there are nested resource types, plan id could be `some-id/image-get/image-get/image-check/...`. To make it clearer, can we change to `some-id/image-get/<type name>/image-get/<other type name>/...`.

Not sure how plan id is used. If the change would lead to complex changes in UI, please just ignore this comment.

aoldershaw replied:
The plan ID doesn't impact too much, so wouldn't be hard to change - that said, it's rarely surfaced to the user (the only place it appears is in the URL when you highlight lines on the build page)

clarafu replied:
I'm not sure if it is necessary to add the type name into the plan_id though because it is just used as a unique identifier. I don't know who would be needing to decode the plan ids, as the plan id of steps in the build plan are just incremented numbers that don't have anything to do with the actual step itself. Did you have anything in mind for using the plan ids?

evanchaoli commented on atc/config.go:
```diff
@@ -296,6 +296,104 @@ func (types ResourceTypes) Without(name string) ResourceTypes {
+		checkPlanID := planID + "/image-check"
+		// don't know the version, need to do a Check before the Get
+		interval := DefaultCheckInterval
+		if checkEvery != nil && !checkEvery.Never {
```
If `checkEvery.Never` is true, should that be a logic error? In that case, should we fail the build?

aoldershaw replied:
`check_every: never` just means "don't periodically check this resource/type", not "never check this resource/type" - so I don't think this needs to result in a failed build.

But that's a good point - if you're setting `check_every: never` on a resource type, it'll now still check it every minute since it's no longer being scheduled via lidar, and every resource check with a custom type has an associated resource type check (that may or may not actually run a container).

Maybe we should just set a really long (practically infinite) interval? We still need to queue up a `check` step, but it shouldn't do anything.

evanchaoli replied:
I think here, `interval` is only used to determine if a new run should run (reaching to check interval). So as Aidan suggested, we can set a really long interval for `never`, so that, as long as a resource type has a version, it will never run auto check again.

clarafu replied:
I do think that is a good idea, though I implemented sort of a rough draft first in order to get your opinion with the `Interval` field on the check plan be a struct that consists of both an `Interval time.Duration` and a `Never bool`, which basically is what the `CheckEvery` struct is when used in the pipeline config. I sort of liked having the `never` value be explicit but if you think the long interval approach would be better or theres a bug in the explicit never bool approach, I'm happy to change it!

evanchaoli commented on atc/lidar/scanner.go:
```diff
@@ -70,26 +72,7 @@ func (s *scanner) scanResources(ctx context.Context, resources []db.Resource, re
-}
-
-func (s *scanner) check(ctx context.Context, checkable db.Checkable, resourceTypes db.ResourceTypes) {
+func (s *scanner) check(ctx context.Context, checkable db.Checkable, resourceTypesMap map[int]db.ResourceTypes) {
```
I agree with @aoldershaw 's comment, empty array is just nil. So here we just just pass in `resourceTypes db.ResourceType` rather than a map.

evanchaoli commented on atc/db/check_factory.go:
```diff
@@ -30,15 +30,15 @@ type Checkable interface {
+	TryCreateCheck(context.Context, Checkable, ResourceTypes, atc.Version, bool, bool) (Build, bool, error)
 	Resources() ([]Resource, error)
-	ResourceTypes() ([]ResourceType, error)
+	ResourceTypes() (map[int]ResourceTypes, error)
```
Maybe rename to `ResourceTypesByPipeline`.

evanchaoli commented on atc/db/check_factory.go:
```diff
@@ -102,23 +84,30 @@ func (c *checkFactory) TryCreateCheck(ctx context.Context, checkable Checkable,
-
-	plan := c.planFactory.NewPlan(checkPlan)
+	deserializedResourceTypes := resourceTypes.Filter(checkable).Deserialize()
+	if deserializedResourceTypes == nil {
```
This check seems not needed. An empty array is equivalent to nil.

evanchaoli commented on atc/db/resource.go:
```diff
@@ -296,22 +296,34 @@ func (r *resource) setResourceConfigScopeInTransaction(tx Tx, scope ResourceConf
-		Source:  sourceDefaults.Merge(r.Source()),
-		Tags:    r.Tags(),
-		Timeout: r.CheckTimeout(),
+type ImagePlanner interface {
```
I doubt if this is a right place to define this interface. Maybe move to a separate file?

evanchaoli commented on atc/db/resource_type.go:
```diff
@@ -48,11 +48,9 @@ type ResourceType interface {
 	SetResourceConfigScope(ResourceConfigScope) error
 
-	CheckPlan(atc.Version, time.Duration, ResourceTypes, atc.Source) atc.CheckPlan
+	CheckPlan(planFactory atc.PlanFactory, imagePlanner ImagePlanner, from atc.Version, interval time.Duration, sourceDefaults atc.Source, manuallyTriggered bool, shallow bool) atc.Plan
```
Should the last parameter `skipIntervalRecursively`?

evanchaoli commented on atc/db/resource_type.go:
```diff
@@ -83,8 +81,8 @@ func (resourceTypes ResourceTypes) Filter(checkable Checkable) ResourceTypes {
-func (resourceTypes ResourceTypes) Deserialize() atc.VersionedResourceTypes {
-	var versionedResourceTypes atc.VersionedResourceTypes
+func (resourceTypes ResourceTypes) Deserialize() atc.ResourceTypes {
+	var atcResourceTypes atc.ResourceTypes
```
If we init it with an empty array, then caller will not need to check nil.

clarafu replied:
I'm not exactly sure what you mean here, I don't think we have anywhere that requires us to check if the resource types is nil or not?

evanchaoli commented:
@clarafu @aoldershaw I went through diffs of this PR, and tested it locally. Now I agree this is a good refactor, making code neater and logic cleaner. Now I see there should be not much conflicts to my PR, I can handle the merge easily. Also, I left a few comments, hope that helps.

clarafu committed:
remove unnecessary test

I think I was trying to test out the looping mechanic and how it would
recheck the last check value, but I did not write the test properly.
Instead, I think it's enough to just test that it checks the last check
value before we try to acquire the lock.

[#6274]

Signed-off-by: Clara Fu <fclara@vmware.com>

clarafu committed:
add test for rechecking last check

Actually I decided that maybe it is valuable to add a test for
refetching the last check value from the db everytime we loop to try to
acquire the lock.

[#6274]

Signed-off-by: Clara Fu <fclara@vmware.com>

clarafu committed:
fix bug with never check every value

Evan pointed out that there was a bug with setting a never value for
check every of a custom type. Prior to this commit, if you set check
every to never on a custom type it would continue to check for the
custom type when it is used as an image.

This commit fixes that were the check step will not run if the user sets
the check every to never for periodic checks. We will still run the
check if it is a manually triggered check.

Also I noticed in the WaitToRun method we were still using the boolean
for build.ManuallyTriggered rather than the SkipInterval value on the
check plan. So I switched the method to use the SkipInterval which is
what it should be using.

[#6274]

Co-authored-by: Esteban Foronda <eforonda@vmware.com>
Signed-off-by: Clara Fu <fclara@vmware.com>

clarafu committed:
fix review comments

Minor fixes and tweaks from review comments

[#6274]

Signed-off-by: Clara Fu <fclara@vmware.com>
